.github/workflows/pip-compile-dev.yml

```diff
@@ -22,39 +22,32 @@ jobs:
           - base-branch: devel
             pr-branch: pip-compile/devel/dev
             nox-args: >-
-              -e 'pip-compile-3.11(formatters)'
-              'pip-compile-3.11(typing)'
-              'pip-compile-3.11(static)'
-              'pip-compile-3.11(spelling)'
-              'pip-compile-3.11(tag)'
+              -e 'pip-compile(formatters)'
+              'pip-compile(typing)'
+              'pip-compile(static)'
+              'pip-compile(spelling)'
+              'pip-compile(tag)'
           - base-branch: stable-2.18
             pr-branch: pip-compile/stable-2.18/dev
             nox-args: >-
-              -e 'pip-compile-3.11(formatters)'
-              'pip-compile-3.11(typing)'
-              'pip-compile-3.11(static)'
-              'pip-compile-3.11(spelling)'
+              -e 'pip-compile(formatters)'
+              'pip-compile(typing)'
+              'pip-compile(static)'
+              'pip-compile(spelling)'
           - base-branch: stable-2.17
             pr-branch: pip-compile/stable-2.17/dev
             nox-args: >-
-              -e 'pip-compile-3.10(formatters)'
-              'pip-compile-3.10(typing)'
```

Comment:
Things like MyPy should typically run against multiple Python versions. Every supported version, ideally. In some projects, I do every other version. Or at least min+max supported.
An invocation w/o args only checks against the current runtime. Deleting this matrix means only executing the check against Python 3.11.

If you end up running it under a single runtime, you can recover the checks of other Pythons by doing multiple invocations.

I typically set the version to check against in the MyPy config to the lowest supported. It'll be picked up when somebody runs it directly.

But then, I add multiple explicit invocations for CI/workflow tooling to cover other branches: https://github.com/ansible/awx-plugins/blob/e7d5733/.pre-commit-config.yaml#L188-L283.

I recommend you do the same here. While this is only dropping some constraint file generation and not the actual execution, I wanted to mention that executing MyPy against different Python versions may require applying additional limitations to some of the requirements just for those invocations.

Reply:
@webknjaz Thanks for the detailed comment about this. I always appreciate getting your input.

My understanding here is that the Python version is declared in the nox session on each branch, such as the one for stable-2.17: `@nox.session(name="pip-compile", python="3.10")` So when the workflow runs each job in the matrix it should use that particular version and not just against Python 3.11?

I feel like I'm missing something from your comment to be honest. I'll try and set this up on my fork and trigger the workflow manually to see what happens.

Reply:
@webknjaz Hey, so it looks like these changes are achieving what we wanted to based on a test run in my fork: https://github.com/oraNod/ansible-documentation/actions/runs/12241705230/job/34147396559

For the devel branch I see this:

```
nox > Creating virtual environment (virtualenv) using python3.11 in .nox/pip-compile-formatters
```

For the stable-2.16 branch, I see this:

```
nox > Creating virtual environment (virtualenv) using python3.10 in .nox/pip-compile-formatters
```

Reply:
I was thinking that maybe we could borrow some inspiration from @gotmax23's clever solution for check mode and update the noxfile like this:

```
@nox.session(name="pip-compile", python=None)
@nox.parametrize(["req"], requirements_files, requirements_files)
def pip_compile(session: nox.Session, req: str, python_version: str = "3.11"):
       if session.python is None:
           session.python = python_version
```

Then we could create a matrix and pass something like `python-version="${{ matrix.python-version }}"` as a nox arg. But that seems unnecessary to me. I still feel like I'm missing something in your comment though.

Reply:
@webknjaz It looks like @samccann merged the backports to stable-2.14-7 so I'm going to go ahead and merge this one too for consistency. Since I checked the workflows in my fork I think it's fine to merge as they do what is expected and fix something that is broken.

I like the idea of the matrix that you suggest but feel like it would be appropriate as a follow on. Let's discuss in the issue and find the best way forward. Cheers.